sign1.go

```diff
@@ -0,0 +1,245 @@
+		return errors.New("Sign1Message signature already has signature bytes")
+	}
+
+	// check algorithm if present
+	skAlg := signer.Algorithm()
+	if alg, err := m.Headers.Protected.Algorithm(); err != nil {
+		if err != ErrAlgorithmNotFound {
+			return err
+		}
+		// `alg` header not present.
+	} else if alg != skAlg {
+		return fmt.Errorf("%w: signer %v: header %v", ErrAlgorithmMismatch, skAlg, alg)
+	}
+
+	// sign the message
+	digest, err := m.digestToBeSigned(skAlg)
+	if err != nil {
+		return err
+	}
+	sig, err := signer.Sign(rand, digest)
+	if err != nil {
+		return err
+	}
+
+	m.Signature = sig
+	return nil
+}
+
+// Verify verifies the signature on the Sign1Message returning nil on success or
+// a suitable error if verification fails.
+//
+// Reference: https://datatracker.ietf.org/doc/html/rfc8152#section-4.4
+func (m *Sign1Message) Verify(verifier Verifier) error {
+	if len(m.Signature) == 0 {
+		return errors.New("Sign1Message has no signature to verify")
+	}
+
+	// check algorithm if present
+	vkAlg := verifier.Algorithm()
+	if alg, err := m.Headers.Protected.Algorithm(); err != nil {
+		if err != ErrAlgorithmNotFound {
+			return err
+		}
+		// `alg` header not present.
+	} else if alg != vkAlg {
+		return fmt.Errorf("%w: verifier %v: header %v", ErrAlgorithmMismatch, vkAlg, alg)
+	}
+
+	// verify the message
+	digest, err := m.digestToBeSigned(vkAlg)
+	if err != nil {
+		return err
+	}
+	return verifier.Verify(digest, m.Signature)
+}
+
+// digestToBeSigned construsts Sig_structure, computes ToBeSigned, and returns
+// the digest of ToBeSigned.
+// If the signing algorithm does not have a hash algorithm associated,
+// ToBeSigned is returned instead.
+//
+// Reference: https://datatracker.ietf.org/doc/html/rfc8152#section-4.4
+func (m *Sign1Message) digestToBeSigned(alg Algorithm) ([]byte, error) {
+	// create a Sig_structure and populate it with the appropriate fields.
+	var protected cbor.RawMessage
+	protected, err := m.Headers.MarshalProtected()
+	if err != nil {
+		return nil, err
+	}
+	external := m.External
+	if external == nil {
+		external = []byte{}
+	}
+	payload := m.Payload
+	if payload == nil {
+		payload = []byte{}
+	}
+	sigStructure := []interface{}{
+		"Signature1", // context
+		protected,    // body_protected
+		external,     // external_aad
+		payload,      // payload
+	}
+
+	// create the value ToBeSigned by encoding the Sig_structure to a byte
+	// string.
+	toBeSigned, err := encMode.Marshal(sigStructure)
+	if err != nil {
+		return nil, err
+	}
+
+	// hash toBeSigned if there is a hash algorithm associated with the signing
+	// algorithm.
+	return alg.computeHash(toBeSigned)
+}
+
+// Sign1 signs a Sign1Message using the provided Signer.
+//
+// This method is a wrapper of `Sign1Message.Sign()`.
+//
+// Reference: https://datatracker.ietf.org/doc/html/rfc8152#section-4.4
+func Sign1(rand io.Reader, signer Signer, protected ProtectedHeader, external, payload []byte) (*Sign1Message, error) {
```

Comment:
nit: std library places the payload before the external data in function signatures. Example: https://pkg.go.dev/crypto/cipher#AEAD. We could follow this convention to avoid mistakes for users using go-cose and std.